<html>
  <head>
    <style type="text/css">
      body {
        background: lightgray;
        font-family: Arial, Helvetica, sans-serif;
      }
      h1 {
        font-size: 2em;
        padding-top: 1em;
        margin: 1em;
      }
      h2 {
        font-size: 1.5em;
        padding-top: 1em;
        margin: 1em;
      }
      p {
        font-size: 1em;
        padding-left: 2em;
        padding-right: 2em;
      }
      table {
        border-collapse: collapse;
        table-layout: fixed;
        margin-left: 2em;
        margin-bottom: 2em;
        padding-bottom: 2em;
        font-size: 0.7em;
      }
      td, th {
        border: 1px solid #999;
        padding: 0.1rem;
      }
      td {
        font-family: "Lucida Console", "Courier New", monospace;
        text-align: right;
        word-wrap: break-word;
      }
      th {
        text-align: left;
      }
      tbody tr:nth-child(odd) {
        background: #eee;
      }
      img {
        width: 100%;
        height: auto;
      }
      .content {
        width: 85% !important;
        margin: auto;
        background: white;
      }
      .subtext {
        font-size: 0.9em;
        padding-left: 2em;
        padding-right: 2em;
      }
      .table-hover tbody tr:hover td, .table-hover tbody tr:hover th {
        background: #ee9;
      }
    </style>
    <title>{{ sample }}</title>
  </head>

  <body>
  <div class="content">
    <!-- Page content -->
    <h1>{{ header }}</h1>

    <img src="{{ summary_plot_fn }}" alt="summary plot">

    <div class="subtext"><b>Figure 1:</b> Summary plot showing the coverage for each sequencing technology and each individual Nanopore amplicon. Regions with a coverage of <{{ threshold_low }} nanopore reads are highlighted in orange. Regions with a coverage of <{{ threshold_limit }} nanopore reads are highlighted in red and the count of afflicted bases is displayed on the top of the figure. Potential variants are displayed below the figure outline. The color of the line indicating each variant's location is green if the potential variant passed the ARITC snv-filter and red otherwise. The text color is green if a given variant is present in the final consensus sequence for this sample. Otherwise it is blue if the variant is known to be associated with any one of the SARS-CoV2 clades defined by Nextstrain. The default text color is black.</div>

    <h2>Nanopore Datasets</h2>

    <p>ARTIC stats</p>

    {{ artic_stats_table }}

    <p>Nanopore runs, sample barcodes and ARTIC guppyplex filter parameters</p>

    {{ nanopore_runs_table }}

    <h2>Potential variants</h2>

    <p>A total of <b>{{ medaka_variant_SNVs|e }}</b> potential SNVs (<b>{{ unique_medaka_variant_SNVs|e }}</b> unique) detected by Medaka variant were passed to longshot for further filtering. Thereof, <b>{{ longshot_SNVs }}</b> met the longshot criteria for being considered potential variants. These are displayed in the table below, in addition to variants that were manually introduced to the final fasta file, if present.</p>

    {{ snv_table }}

    <h2>Nextstrain clade assignment</h2>

    <p>Assignment produced by Nextstrain's assign_clade.py script: <b>{{ clade_assignment|e }}</b>, parent clade: {{ parent_clade|e }}</p>

    {% if final %}
    <h2>N-masked regions</h2>

    <p>The final consensus sequence is <b>{{ consensus_length }}</b> bases long. In total, <b>{{ masked_bases }}</b> bases in the sequence are masked (N in consensus sequence). The following table lists all masked regions.</p>

    {{ masked_regions_table }}
    {% endif %}

    <h2>Software versions</h2>

    {{ software_versions_table }}

  </div>
  </body>
</html>